src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'clear':
+        clearProperties(stateCopy);
+        stateHistory.push({ ...stateCopy });
+        break;
+    }
```

Comment:
[CODE STYLE]: switch/case should always have default case for error handling. Add a default case to handle unknown action types.